warehouse/transformer/benchmark/main.go

```diff
@@ -0,0 +1,143 @@
+package main
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"log"
+	"os"
+	"os/signal"
+	"strconv"
+	"strings"
+	"syscall"
+	"time"
+
+	"github.com/google/uuid"
+	"golang.org/x/sync/errgroup"
+
+	"github.com/rudderlabs/rudder-go-kit/config"
+	"github.com/rudderlabs/rudder-go-kit/logger"
+	"github.com/rudderlabs/rudder-go-kit/profiler"
+	"github.com/rudderlabs/rudder-go-kit/stats"
+
+	"github.com/rudderlabs/rudder-server/jsonrs"
+	ptrans "github.com/rudderlabs/rudder-server/processor/transformer"
+	"github.com/rudderlabs/rudder-server/processor/types"
+	"github.com/rudderlabs/rudder-server/utils/timeutil"
+	wtrans "github.com/rudderlabs/rudder-server/warehouse/transformer"
+)
+
+const (
+	processorTransformer = "PROCESSOR_TRANSFORMER"
+	warehouseTransformer = "WAREHOUSE_TRANSFORMER"
+)
+
+func main() {
+	ctx, cancel := signal.NotifyContext(context.Background(), syscall.SIGKILL, syscall.SIGINT, syscall.SIGTERM)
+	defer cancel()
+
+	if err := run(ctx); err != nil {
+		log.Fatal(err)
+	}
+}
+
+func run(ctx context.Context) error {
+	sampleEvent := mustString("SAMPLE_EVENT")
+	eventsInBatch, batchSize := mustInt("EVENTS_IN_BATCH"), mustInt("BATCH_SIZE")
+	iterations := mustInt("NUM_OF_ITERATIONS")
+	mode := mustString("MODE")
+
+	if eventsInBatch < batchSize {
+		return errors.New("eventsInBatch must be greater than or equal to batchSize")
+	}
+
+	runID := uuid.New().String()
+	conf := config.Default
+
+	logFactory := logger.Default
+	l := logFactory.NewLogger().Child("warehouse-transformer-benchmark").Child(mode)
+
+	clientEvents := make([]types.TransformerEvent, eventsInBatch)
+	for i := 0; i < len(clientEvents); i++ {
+		var transformerEvent types.TransformerEvent
+		err := jsonrs.Unmarshal([]byte(sampleEvent), &transformerEvent)
```

Comment:
Since value of `sampleEvent` is fixed, it can be unmarshalled only once

Reply:
We can't use a single `transformerEvent`, because it contains `SingularEventT` which is a **map** because **maps** in Go are reference types. One option is to just do Unmarshal everytime. Other option is to do Unmarshal once but then later do deep copy using something like:
```GO
func deepCopy(src, dest interface{}) error {
	buf, err := jsonrs.Marshal(src)
	if err != nil {
		return err
	}
	return jsonrs.Unmarshal(buf, dest)
}
```

In favour of the first approach.